src/tasks/getsubscriptiondetails/taskgetsubscriptiondetails.cpp

```diff
@@ -0,0 +1,120 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "taskgetsubscriptiondetails.h"
+#include "authenticationlistener.h"
+#include "authenticationinapp/authenticationinappsession.h"
+#include "errorhandler.h"
+#include "leakdetector.h"
+#include "logger.h"
+#include "mozillavpn.h"
+#include "networkrequest.h"
+
+namespace {
+Logger logger(LOG_MAIN, "TaskGetSubscriptionDetails");
+}  // anonymous namespace
+
+TaskGetSubscriptionDetails::TaskGetSubscriptionDetails(
+    const QString& emailAddress)
+    : Task("TaskGetSubscriptionDetails"), m_emailAddress(emailAddress) {
+  MVPN_COUNT_CTOR(TaskGetSubscriptionDetails);
+}
+
+TaskGetSubscriptionDetails::~TaskGetSubscriptionDetails() {
+  MVPN_COUNT_DTOR(TaskGetSubscriptionDetails);
+}
+
+void TaskGetSubscriptionDetails::run() {
+  logger.debug() << "run";
+
+  NetworkRequest* request =
+      NetworkRequest::createForGetSubscriptionDetails(this);
+
+  connect(
+      request, &NetworkRequest::requestFailed, this,
+      [this](QNetworkReply::NetworkError error, const QByteArray&) {
+        logger.error() << "Get subscription details failed" << error;
+
+        // Network request failed after authentication for a second time
+        if (m_authenticationInAppSession) {
+          MozillaVPN::instance()->errorHandle(ErrorHandler::toErrorType(error));
+          emit failed();
+          m_authenticationInAppSession->terminate();
+          return;
+        }
+
+        // User needs to (re)authenticate
+        if (error == QNetworkReply::AuthenticationRequiredError) {
+          emit needsAuthentication();
+          initAuthentication();
+          return;
+        }
+
+        MozillaVPN::instance()->errorHandle(ErrorHandler::toErrorType(error));
+        emit failed();
+        emit completed();
```

Comment:
why both? :)

Reply:
In order to be able to explicitly know that the `TaskGetSubscriptionDetails` failed without additional checks. I thought it might be useful, but it’s currently only used for debugging purposes [here](https://github.com/mozilla-mobile/mozilla-vpn-client/pull/3606/files#diff-1fba09d53eb0acfed1f8c589ae89ce8326577b6160f5b5976ea0b2557755ccc3R44).